# Unveiling Mechanisms in Vaginal Microbiome Dynamics through Model-Driven Analysis

In this study, we analyzed metagenomic data obtained from human vaginal swabs to generate in silico predictions of BV-associated bacterial metabolic interactions via genome-scale metabolic network reconstructions (GENREs). We grew several of the most common co-occurring bacteria (Prevotella amnii, Prevotella buccalis, Hoylesella timonensis, Lactobacillus iners, Fannyhessea vaginae, and Aerrococcus christenssii) on the spent media of Gardnerella species and performed metabolomics to identify potential mechanisms of metabolic interaction.

## [Analysis](https://github.com/lrd3uu/bacterialvaginosis_interactions/tree/main/Analysis)
Followup analysis of the simulations output. 

## [Simulations](https://github.com/lrd3uu/bacterialvaginosis_interactions/tree/main/Simulations)
Code required to generate and run the 15,000 simulations analyzed in the aforementioned study

## [Metabolomics](https://github.com/lrd3uu/bacterialvaginosis_interactions/tree/main/Metabolomics)

The processed metabolomics data files used for the data analysis portion. The raw files for the [method development](https://www.dropbox.com/scl/fo/yz5gpej71avgetdl2aa26/h?rlkey=m9wvvi0ep7y8qc00cnmnxc5jh&dl=0), [samples and standard curves](https://www.dropbox.com/scl/fo/wlsa6q8h8ruk6xkc2fiaz/h?rlkey=17krvbcmsbnfjajng3jzu4ncd&dl=0), as well as [metabolomics data](https://www.dropbox.com/scl/fo/dsaejf46kphv5vzhc67j0/h?rlkey=31rfyt9nam8m91i9py38zpuvu&dl=0) from our untargeted metabolomics analysis performed by the UVA mass spectrometry team



